mailbox/cassandra/src/main/java/org/apache/james/mailbox/cassandra/mail/CassandraMessageMapper.java

```diff
@@ -320,37 +324,45 @@ private boolean updateFlags(MailboxMessage message, long oldModSeq) {
                 .modSeq(message.getModSeq())
                 .flags(message.createFlags())
                 .build();
-        CompletableFuture<Boolean> messageIdFuture = messageIdDAO.updateMetadata(composedMessageIdWithMetaData, oldModSeq);
-        CompletableFuture<Boolean> imapUidFuture = imapUidDAO.updateMetadata(composedMessageIdWithMetaData, oldModSeq);
-        return messageIdFuture.join() && imapUidFuture.join();
+        return imapUidDAO.updateMetadata(composedMessageIdWithMetaData, oldModSeq)
+            .thenCompose(success -> {
+                if (success) {
+                    return messageIdDAO.updateMetadata(composedMessageIdWithMetaData).thenApply(any -> success);
+                }
+                return CompletableFuture.completedFuture(success);
+            })
+            .join();
     }
 
     private Optional<UpdatedFlags> handleRetries(Mailbox mailbox, FlagsUpdateCalculator flagUpdateCalculator, MailboxMessage message) {
         try {
             return Optional.of(
                 new FunctionRunnerWithRetry(maxRetries)
                     .executeAndRetrieveObject(() -> retryMessageFlagsUpdate(mailbox,
-                            ComposedMessageIdWithMetaData.builder()
-                                .composedMessageId(new ComposedMessageId(mailbox.getMailboxId(), message.getMessageId(), message.getUid()))
-                                .modSeq(message.getModSeq())
-                                .flags(message.createFlags())
-                            .build(),
+                            message.getMessageId(),
                             flagUpdateCalculator)));
         } catch (MessageDeletedDuringFlagsUpdateException e) {
             mailboxSession.getLog().warn(e.getMessage());
             return Optional.empty();
+        } catch (MailboxDeleteDuringUpdateException e) {
+            LOGGER.info("Mailbox {} was deleted during flag update", mailbox.getMailboxId());
+            return Optional.empty();
         } catch (Exception e) {
             throw Throwables.propagate(e);
         }
     }
 
-    private Optional<UpdatedFlags> retryMessageFlagsUpdate(Mailbox mailbox, ComposedMessageIdWithMetaData composedMessageIdWithMetaData, FlagsUpdateCalculator flagUpdateCalculator) {
-        CassandraId mailboxId = (CassandraId) mailbox.getMailboxId();
+    private Optional<UpdatedFlags> retryMessageFlagsUpdate(Mailbox mailbox, MessageId messageId, FlagsUpdateCalculator flagUpdateCalculator) {
+        CassandraId cassandraId = (CassandraId) mailbox.getMailboxId();
+        ComposedMessageIdWithMetaData composedMessageIdWithMetaData = imapUidDAO.retrieve((CassandraMessageId) messageId, Optional.of(cassandraId))
```

Comment:
looks like a snippet copy/pasted two or more time, isn't it ? Maybe it deserves a method in ImapUidDao ?